src/core/service.c

```diff
@@ -3472,8 +3472,18 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
 
                         switch (s->state) {
 
-                        case SERVICE_START_POST:
                         case SERVICE_RELOAD:
+                                /* The service has exited but it's still in
+                                 * reloading state, we can't expect the reload
+                                 * to complete so end the control process.
+                                 */
```

Comment:
so i figure you want to check here if the control pid is still valid, and if so, immediately enter the stop phase.

```c
case SERVICE_START_POST:
case SERVICE_RELOAD:

        if (control_pid_good(s) <= 0) /* If from here on neither a main nor a control process remain running, then the current state can never exit cleanly, hence immediately terminate the service */
                service_enter_stop(s, SERVICE_SUCCESS);

        /* Otherwise we need to wait until the operation is complete. */
         break;
```

Reply:
Thanks, pushed new changes. I've propagated the original ServiceResult "f" here though instead of passing SERVICE_SUCCESS since this seems to more accurately capture what the state was at the time of the service exiting.